Make Marshaler internal and rename

jonlawlor opened the pull request:
I’ve changed `MarshalParamers` and `UnmarshalParameters` to
`getParameters` and `setParameters`, respectively.  I’ve also modified
the testConjugateUpdate function to be a table driven test, and perform
more comparisons.  It turns out that by modifying its signature, it
wasn’t necessary to use the new `parameterMutator` interface, although
that will still be useful for other serialization.

I implemented `setParameters` and `getParameters` as discussed
previously, but it would also be possible to write a single 
`Parameters([]Parameter) []Parameter` which would act as both set and
get in the same manner as `runtime.GOMAXPROCS`.

This follows up #31, PTAL @btracey, @kortschak

btracey commented on dist/exponential.go:
```diff
@@ -229,8 +218,26 @@ func (e Exponential) Survival(x float64) float64 {
+}
+
+// getParameters implements the parameterMutator interface
+func (e Exponential) getParameters(p []Parameter) []Parameter {
```
Shouldn't this just be "parameters"? Go generally doesn't use "Get" unless otherwise necessary.

kortschak replied:
Unless it collides with an existing name `parameters`. In which case add to TODO to name it `Parameters` when it is exported.

btracey commented on dist/general.go:
```diff
@@ -6,20 +6,17 @@ type Parameter struct {
-// distribution.
-// Marshal and Unmarshal are to be used as a pair, users should not attempt to construct
+// A parameterMutator is a type that can both get and set the parameters of a distribution.
+// Typically, users should modify distributions using the fields of the specifc distribution.
```
I don't think this "users" comment is necessary now that it's an internal function. There aren't any users involved.

kortschak replied:
Spelling of "specific" if kept.

btracey commented on dist/general_test.go:
```diff
@@ -64,126 +61,73 @@ func testDistributionProbs(t *testing.T, dist UniProbDist, name string, pts []un
-	weights := make([]float64, len(samps))
-	for i := range weights {
-		weights[i] = rand.Float64()
+func testConjugateUpdate(t *testing.T, newFittable func() ConjugateUpdater, isEqual func(ConjugateUpdater, ConjugateUpdater) bool) {
```
I don't like the second argument here, the reflection (in exponential for example) isn't necessary. We we can use getParameters here instead and compare the parameters directly.

kortschak commented on dist/general.go:
```diff
@@ -6,20 +6,17 @@ type Parameter struct {
+// appropriate size will be allocated.
+// setParameters will panic if the names of the parameters do not match. setParameters tests names
+// in the same order as they were created in getParameters.
+type parameterMutator interface {
```
I don't like this name, but I can't think of anything better.

jonlawlor replied:
Could call it `Parametric`, `Parameterizable`, or even split it into `ParameterGetter` and `ParameterSetter`.

kortschak commented:
>it would also be possible to write a single
Parameters([]Parameter) []Parameter which would act as both set and
get in the same manner as runtime.GOMAXPROCS.

That is not possible in this case, since you may be passing in a non-zero `[]Parameter` merely to save the allocation.

The reordering of source is a little confusing in the diff here (Variance moves so adds diff lines that aren't relevant AFAICS).

kortschak commented:
>It turns out that by modifying its signature, it
wasn’t necessary to use the new `parameterMutator` interface

Then drop it I think and add it when we need it.

btracey commented on dist/general_test.go:
```diff
@@ -64,126 +61,73 @@ func testDistributionProbs(t *testing.T, dist UniProbDist, name string, pts []un
+		samps   []float64
+		weights []float64
+	}{
+		{
```
Could you recombine the nil case and the ones case? Maybe with a "if weights == nil, test with ones". Those two answers should be the same, which is otherwise untested. I found bugs in my initial implementations because of checking that case.

jonlawlor commented:
I reordered it because it seemed like the style was to have unexported functions at the bottom of files.  If you want I can move it back.

Otherwise I've made the requested changes, PTAL @kortschak, @btracey.

kortschak commented:
Yes, please revert the order change; that can be part/all of another PR.

I still don't like that we have an interface that's only barely used. Can we drop `parameterized` and use the explicit methods in `ConjugateUpdater`. In the long run I would prefer that we use (when they are exported), `ParameterSetter` and `ParameterGetter` (still ugh, but probably the best available) as separate interfaces. Clients can define their own parameterized (without the `NumParameters`, or with, as they need).

jonlawlor commented:
Alright, I've moved `Variance` back and also removed the `parameterized` interface.

I think that the `parameters` and `setParameters` functions are going to be much slower than direct field access.  If we need to compare distributions, a set of hypothesis test functions would probably be faster.  In the worst case, we could use the delegate approach I was using previously in the `testConjugateUpdate` function.

kortschak commented:
Would you remove the comment references to parameterized since it no longer exists.

jonlawlor commented:
Done

jonlawlor commented:
I think this is ready to merge, PTAL @btracey, @kortschak

jonlawlor commented:
Also, on my list of things to do:

* add `CorrelationMatrix` to stat
* add `LogNormal` to dist
* add `Student` to dist (name ok?)

kortschak commented:
LGTM
